server/src/main/java/org/candlepin/async/JobStateManagementException.java

```diff
@@ -0,0 +1,273 @@
+     */
+    public JobStateManagementException(AsyncJobStatus status, JobState initialState, JobState intendedState,
+        String message, boolean terminal) {
+
+        this(status, initialState, intendedState, null, null, false);
+    }
+
+    /**
+     * Constructs a new exception with the specified cause and a detail message of
+     * <tt>(cause == null ? null : cause.toString())</tt> (which typically contains the and
+     * detail message of cause). This constructor is useful for exceptions that are little more
+     * than wrappers for other throwables (for example, PrivilegedActionException).
+     *
+     * @param status
+     *  The AsyncJobStatus instance being updated at the time of failure
+     *
+     * @param initialState
+     *  The initial state of the job status before any changes were made leading to the exception
+     *
+     * @param intendedState
+     *  The intended state the job status was transitioning to at the time of the exception
+     *
+     * @param cause
+     *  the cause (which is saved for later retrieval by the Throwable.getCause() method). A null
+     *  value is permitted, and indicates that the cause is nonexistent or unknown.
+     *
+     * @param terminal
+     *  whether or not the exception is terminal or non-recoverable and the job should not be
+     *  retried and any associated job messages should be discarded
+     */
+    public JobStateManagementException(AsyncJobStatus status, JobState initialState, JobState intendedState,
+        Throwable cause, boolean terminal) {
+
+        this(status, initialState, intendedState, null, null, false);
+    }
+
+    /**
+     * Constructs a new exception with the specified detail message and cause.
+     * <p></p>
+     * Note that the detail message associated with cause is not automatically incorporated in this
+     * exception's detail message.
+     *
+     * @param status
+     *  The AsyncJobStatus instance being updated at the time of failure
+     *
+     * @param initialState
+     *  The initial state of the job status before any changes were made leading to the exception
+     *
+     * @param intendedState
+     *  The intended state the job status was transitioning to at the time of the exception
+     *
+     * @param message
+     *  the detail message. The detail message is saved for later retrieval by the getMessage()
+     *  method.
+     *
+     * @param cause
+     *  the cause (which is saved for later retrieval by the Throwable.getCause() method). A null
+     *  value is permitted, and indicates that the cause is nonexistent or unknown.
+     *
+     * @param terminal
+     *  whether or not the exception is terminal or non-recoverable and the job should not be
+     *  retried and any associated job messages should be discarded
+     */
+    public JobStateManagementException(AsyncJobStatus status, JobState initialState, JobState intendedState,
+        String message, Throwable cause, boolean terminal) {
```

Comment:
It's exactly because they're optional that there's so many constructors. After the required status bits, the remaining parameters can be provided in any combination; in line with the "standard" exception constructor arrangement.

I don't really want to make a builder for something that's not intended to be used outside of its current use anyway.